File: LeetCode/1433. Check If a String Can Break Another String/solution.py
class Solution:
    def check(self, d1, d2):
        s = 0
        for c in 'abcdefghijklmnopqrstuvwxyz':
            s += d1[c] - d2[c]
            if s < 0:
                return False
        return True
    
    def checkIfCanBreak(self, s1: str, s2: str) -> bool:
        d1 = collections.Counter(s1)
        d2 = collections.Counter(s2)
        return self.check(d1, d2) | self.check(d2, d1)


class Solution:
    def checkIfCanBreak(self, s1: str, s2: str) -> bool:
        def cmp(A,B):
            for a, b in zip(A,B):
                if a < b:
                    return False
            return True

        s1 = ''.join(sorted(list(s1))[::-1])
        s2 = ''.join(sorted(list(s2))[::-1])
        return cmp(s1,s2) or cmp(s2,s1)

File: LeetCode/1433. Check If a String Can Break Another String/solution.cpp
// class Solution {
// public:
//     bool great(string s1, string s2){
//         int n = s1.length();
//         for (int i = 0; i < n; i++)
//             if (s1[i] < s2[i])
//                 return false;
//         return true;
//     }
//     bool checkIfCanBreak(string s1, string s2) {
//         sort(s1.begin(), s1.end());
//         sort(s2.begin(), s2.end());
//         return great(s1,s2) || great(s2,s1);
//     }
// };


class Solution {
public:
    bool check(string s1, string s2) {
        int book1[27] = {0}, book2[27] = {0};
        for (auto p: s1) {
            book1[p - 'a']++;
        }
        // 检查s1能不能打击s2
        for(auto p: s2) {
            // 找一个最省力的s1中元素
            for (int i = p - 'a'; i <= 'z' - 'a'; i++) {
                if (book1[i] > 0) {
                    book1[i]--; // 用掉一个元素
                    break;
                }
                if (i == 'z' - 'a') {
                    return false; // 出错！
                }
            }
        }
        return true;
    }
    bool checkIfCanBreak(string s1, string s2) {
        return check(s1, s2) || check(s2, s1);
    }
};
